hca/parser.py

```diff
@@ -15,36 +23,331 @@ def make_name(http_method, path_split):
+                options[param_name]["required_for"].append(path)
+
+
+def _get_arg_type(arg_type_string):
+    argtype = None
+    if arg_type_string == "integer":
+        argtype = int
+    elif arg_type_string == "boolean":
+        argtype = bool
+    elif arg_type_string == "number":
+        argtype = float
+    return argtype
+
+
+def _get_action(arg_type_string):
+    if arg_type_string == "object":
+        return AddObject
+    return "store"
+
+
+class AddObject(argparse.Action):
+    """Object to parse json objects."""
+
+    def __init__(self, option_strings, dest, nargs=None, **kwargs):
+        super(AddObject, self).__init__(option_strings, dest, **kwargs)
+    def __call__(self, parser, namespace, values, option_string=None):
+        setattr(namespace, self.dest, json.loads(values))
+
+
+def _add_positional_args(subparser, endpoint_info):
+    for positional_arg in endpoint_info["positional"]:
+        h = create_help_statement(positional_arg)
+
+        argtype = _get_arg_type(positional_arg["type"])
+
+        subparser.add_argument(
+            positional_arg["argument"],
+            nargs=None if positional_arg["required"] else "?",
+            help=h,
+            type=argtype
+        )
+
+
+def _add_optional_args(subparser, endpoint_info):
+    for (optional_name, optional_data) in endpoint_info["options"].items():
+        h = create_help_statement(optional_data)
+
+        argtype = _get_arg_type(optional_data["type"])
+        actiontype = _get_action(optional_data["type"])
+
+        optional_name_with_dashes = optional_name.replace("_", "-")
+        subparser.add_argument(
+            "--" + optional_name_with_dashes,
+            dest=optional_name,  # So we can send correctly formatted objects to api.
+            metavar=optional_data["metavar"],
+            required=optional_data["required"],
+            nargs="+" if optional_data['array'] else None,
+            help=h,
+            type=argtype,
+            action=actiontype
+        )
+
+
 def get_parser(spec):
-    """Return a cli parser that adheres to the api defined in api_spec.json."""
+    """
+    Return an argparse parser and a dict containing important argument details.
+
+    :param spec: The swagger api specification used to generate parser.
+    :return parser: The parser configured to the given api spec.
+    :return param_holders: Important parameter information for each endpoint used in
+                           parsing. Basic format is:
+                           {
+                                <endpoint-name>: {
+                                    description: str,
+                                    options: {
+                                        <option-arg-name> {
+                                            array: bool,
+                                            description: str,
+                                            format: str,
+                                            hierarchy: list,
```

Comment:
It's described fully in `define_api.py`. I'll add a short description next to this